From 6270d5279c78725e74a8ed88a60d67bd0296b957 Mon Sep 17 00:00:00 2001
From: nkk71 <nkk71x@gmail.com>
Date: Fri, 19 Jun 2015 18:25:08 +0300
Subject: [PATCH] fix SELinux "unlabled" /data/.layout_version (using init.rc
 import)

---
 rom_quirks.c | 70 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 70 insertions(+)

diff --git a/rom_quirks.c b/rom_quirks.c
index 92f0fc7..1ab8630 100644
--- a/rom_quirks.c
+++ b/rom_quirks.c
@@ -104,6 +104,68 @@ static void inject_file_contexts(void)
     fclose(f);
 }
 
+
+static void inject_initrc_restorecon_layoutversion(void)
+{
+    FILE *f;
+    char line[512];
+
+    f = fopen("/init.rc", "re");
+    if(!f)
+    {
+        ERROR("Failed to open /init.rc!");
+        return;
+    }
+
+    // check if import already exits
+    while(fgets(line, sizeof(line), f))
+    {
+        if(strstartswith(line, "import /init.layout_version.rc"))
+        {
+            INFO("/init.rc has been already injected.");
+            fclose(f);
+            return;
+        }
+    }
+
+    fclose(f);
+
+
+    // create the init.layout_version.rc file which has the
+    // restorecon command
+    INFO("Create /init.layout_version.rc");
+    f = fopen("/init.layout_version.rc", "w");
+    if(!f)
+    {
+        ERROR("Failed to create /init.layout_version.rc!");
+        return;
+    }
+
+    fputs("on early-init\n"
+          "    # Set the security context of /data/.layout_version if present.\n"
+          "    restorecon /data/.layout_version\n",
+        f);
+    fclose(f);
+    chmod("/init.layout_version.rc", 0750);
+
+
+    // go ahead add the import
+    INFO("Injecting /init.rc");
+    f = fopen("/init.rc", "ae");
+    if(!f)
+    {
+        ERROR("Failed to open /init.rc for appending!");
+        return;
+    }
+
+    fputs("\n"
+        "# fix for SELinux unlabled layout_version\n"
+        "import /init.layout_version.rc\n",
+        f);
+    fclose(f);
+}
+
+
 void rom_quirks_on_initrd_finalized(void)
 {
     // walk over all _regular_ files in /
@@ -127,6 +189,14 @@ void rom_quirks_on_initrd_finalized(void)
             if(strcmp(dt->d_name, "file_contexts") == 0)
                 inject_file_contexts();
 
+
+            // "/data/.layout_version" is created without proper SELinux
+            // context during "multirom_create_media_link" function
+            // add a "restorecon /data/.layout_version" to init.rc
+            if(strcmp(dt->d_name, "init.rc") == 0)
+                inject_initrc_restorecon_layoutversion();
+
+
             // franco.Kernel includes script init.fk.sh which remounts /system as read only
             // comment out lines with mount and /system in all .sh scripts in /
             if(strendswith(dt->d_name, ".sh"))
diff --git a/version.h b/version.h
index 245eb78..b9d2f83 100644
--- a/version.h
+++ b/version.h
@@ -21,5 +21,5 @@
     #define VERSION_TRAMPOLINE 26
 
     // For device-specific fixes. Use letters, the version will then be like "12a"
-    #define VERSION_DEV_FIX ""
+    #define VERSION_DEV_FIX "a"
 #endif
