diff --git a/Android.mk b/Android.mk
index 5d32622..426beb4 100644
--- a/Android.mk
+++ b/Android.mk
@@ -21,9 +21,14 @@ LOCAL_MODULE                  := mrom_kexec_static
 LOCAL_MODULE_TAGS             := optional
 LOCAL_C_INCLUDES              := $(LOCAL_PATH)/include \
                                  $(LOCAL_PATH)/util_lib/include \
-                                 $(LOCAL_PATH)/kexec/arch/arm/include \
                                  $(LOCAL_PATH)/kexec/libfdt \
+                                 $(LOCAL_PATH)/kexec \
                                  external/zlib
+ifeq ($(TARGET_ARCH),arm64)
+LOCAL_C_INCLUDES              += $(LOCAL_PATH)/kexec/arch/arm64/include
+else
+LOCAL_C_INCLUDES              += $(LOCAL_PATH)/kexec/arch/arm/include
+endif
 LOCAL_SRC_FILES               := kexec/kexec.c kexec/ifdown.c \
                                  kexec/kexec-elf.c kexec/kexec-elf-exec.c \
                                  kexec/kexec-elf-core.c \
@@ -34,24 +39,34 @@ LOCAL_SRC_FILES               := kexec/kexec.c kexec/ifdown.c \
                                  kexec/crashdump.c kexec/crashdump-xen.c \
                                  kexec/phys_arch.c kexec/lzma.c \
                                  kexec/zlib.c kexec/proc_iomem.c \
-                                 kexec/virt_to_phys.c \
-                                 kexec/arch/arm/phys_to_virt.c \
-                                 kexec/add_segment.c kexec/add_buffer.c \
-                                 kexec/arch_reuse_initrd.c \
+                                 kexec/add_buffer.c \
                                  kexec/arch_init.c \
+                                 kexec/kexec-uImage.c kexec/purgatory.c \
+                                 kexec/fs2dt.c \
+                                 kexec/dt-ops.c \
+                                 kexec/libfdt/fdt.c kexec/libfdt/fdt_ro.c \
+                                 kexec/libfdt/fdt_rw.c kexec/libfdt/fdt_strerror.c \
+                                 kexec/libfdt/fdt_sw.c kexec/libfdt/fdt_wip.c
+ifeq ($(TARGET_ARCH),arm64)
+LOCAL_SRC_FILES               += kexec/arch/arm64/kexec-elf-arm64.c \
+                                 kexec/arch/arm64/kexec-image-arm64.c \
+                                 kexec/arch/arm64/kexec-arm64.c \
+                                 kexec/arch/arm64/crashdump-arm64.c
+else
+LOCAL_SRC_FILES               += kexec/arch/arm/phys_to_virt.c \
                                  kexec/arch/arm/kexec-elf-rel-arm.c \
+                                 kexec/add_segment.c \
+                                 kexec/arch_reuse_initrd.c \
+                                 kexec/virt_to_phys.c \
                                  kexec/arch/arm/kexec-zImage-arm.c \
                                  kexec/arch/arm/kexec-uImage-arm.c \
                                  kexec/arch/arm/kexec-arm.c \
                                  kexec/arch/arm/mach.c \
                                  kexec/arch/arm/mach-hammerhead.c \
                                  kexec/arch/arm/mach-m8.c \
                                  kexec/arch/arm/mach-shamu.c \
-                                 kexec/arch/arm/crashdump-arm.c \
-                                 kexec/kexec-uImage.c kexec/purgatory.c \
-                                 kexec/libfdt/fdt.c kexec/libfdt/fdt_ro.c \
-                                 kexec/libfdt/fdt_rw.c kexec/libfdt/fdt_strerror.c \
-                                 kexec/libfdt/fdt_sw.c kexec/libfdt/fdt_wip.c
+                                 kexec/arch/arm/crashdump-arm.c
+endif
 LOCAL_FORCE_STATIC_EXECUTABLE := true
 LOCAL_STATIC_LIBRARIES        := mrom_libutil_kt libz libc
 LOCAL_MODULE_PATH := $(TARGET_OUT_OPTIONAL_EXECUTABLES)
diff --git a/kexec/arch/arm/include/kexec-fs2dt.h b/kexec/arch/arm/include/kexec-fs2dt.h
new file mode 100644
index 0000000..04e6829
--- /dev/null
+++ b/kexec/arch/arm/include/kexec-fs2dt.h
@@ -0,0 +1,7 @@
+#ifndef KEXEC_FS2DT_H
+#define KEXEC_FS2DT_H
+
+#include "arch/arm/kexec-arm.h"
+#include "arch/arm/crashdump-arm.h"
+
+#endif /* KEXEC_FS2DT_H */
diff --git a/kexec/arch/arm64/crashdump-arm64.c b/kexec/arch/arm64/crashdump-arm64.c
index a929b9e..03378da 100644
--- a/kexec/arch/arm64/crashdump-arm64.c
+++ b/kexec/arch/arm64/crashdump-arm64.c
@@ -215,7 +215,7 @@ static void dump_crash_ranges(void)
 {
 	int i;
 
-	if (!kexec_debug)
+	if (true) // don't need debugging
 		return;
 
 	dbgprintf("%s: kernel: %016llx - %016llx (%ld MiB)\n", __func__,
diff --git a/kexec/arch/arm64/include/kexec-fs2dt.h b/kexec/arch/arm64/include/kexec-fs2dt.h
new file mode 100644
index 0000000..0599ef0
--- /dev/null
+++ b/kexec/arch/arm64/include/kexec-fs2dt.h
@@ -0,0 +1,7 @@
+#ifndef KEXEC_FS2DT_H
+#define KEXEC_FS2DT_H
+
+#include "arch/arm64/kexec-arm64.h"
+#include "arch/arm64/crashdump-arm64.h"
+
+#endif /* KEXEC_FS2DT_H */
diff --git a/kexec/arch/arm64/kexec-arm64.c b/kexec/arch/arm64/kexec-arm64.c
index de9f962..ae1ff51 100644
--- a/kexec/arch/arm64/kexec-arm64.c
+++ b/kexec/arch/arm64/kexec-arm64.c
@@ -109,7 +109,6 @@ int arch_process_options(int argc, char **argv)
 	}
 
 	arm64_opts.command_line = concat_cmdline(tmp_cmdline, append);
-	kexec_debug = 1; // FIXME: for debugging only.
 
 	dbgprintf("%s:%d: command_line: %s\n", __func__, __LINE__,
 		arm64_opts.command_line);
diff --git a/kexec/fs2dt.h b/kexec/fs2dt.h
index cd85aee..f0ff002 100644
--- a/kexec/fs2dt.h
+++ b/kexec/fs2dt.h
@@ -1,6 +1,8 @@
 #ifndef FS2DT_H
 #define FS2DT_H
 
+#include <kexec-fs2dt.h>
+
 #if (BOOT_BLOCK_VERSION != 2 && BOOT_BLOCK_VERSION != 17)
 #error Please add or correct definition of BOOT_BLOCK_VERSION
 #endif
diff --git a/kexec/arch/arm64/crashdump-arm64.h b/kexec/arch/arm64/crashdump-arm64.h
index a7b67d1..d01a8b0 100644
--- a/kexec/arch/arm64/crashdump-arm64.h
+++ b/kexec/arch/arm64/crashdump-arm64.h
@@ -16,6 +16,11 @@
 
 #define CRASH_MAX_MEMORY_RANGES	32
 
+struct memory_ranges {
+        unsigned int size;
+        struct memory_range *ranges;
+};
+
 extern struct memory_ranges usablemem_rgns;
 extern struct memory_range crash_reserved_mem;
 
