diff --git a/multirom.c b/multirom.c
index 9da3e08..0a02b80 100644
--- a/multirom.c
+++ b/multirom.c
@@ -1645,9 +1645,9 @@ int multirom_load_kexec(struct multirom_status *s, struct multirom_rom *rom)
 
     kexec_init(&kexec, kexec_path);
     kexec_add_arg(&kexec, "--mem-min="MR_KEXEC_MEM_MIN);
-#ifdef MR_KEXEC_DTB
-    kexec_add_arg_prefix(&kexec, "--boardname=", TARGET_DEVICE);
-#endif
+//#ifdef MR_KEXEC_DTB
+//    kexec_add_arg_prefix(&kexec, "--boardname=", TARGET_DEVICE);
+//#endif
 
     switch(rom->type)
     {
@@ -1717,8 +1717,8 @@ int multirom_fill_kexec_android(struct multirom_status *s, struct multirom_rom *
 #ifdef MR_KEXEC_DTB
     if(libbootimg_dump_dtb(&img, "/dtb.img") >= 0)
         kexec_add_arg(kexec, "--dtb=/dtb.img");
-    else
-        kexec_add_arg(kexec, "--dtb");
+//    else
+//        kexec_add_arg(kexec, "--dtb");
 #endif
 
     char cmdline[1536];
