diff --git a/kexec-tools b/kexec-tools
index 1945c66..b17408c 160000
--- a/kexec-tools
+++ b/kexec-tools
@@ -1 +1 @@
-Subproject commit 1945c661d8b117297d3ac44ea071556298c31ec3
+Subproject commit b17408ca57403d66dc8bf7e0049bbd6bbfac8565
diff --git a/trampoline/trampoline.c b/trampoline/trampoline.c
index b9a10e9..faf555b 100644
--- a/trampoline/trampoline.c
+++ b/trampoline/trampoline.c
@@ -202,9 +202,9 @@ static int mount_and_run(struct fstab *fstab)
         return -1;
     }
 
-    adb_init(path_multirom);
+    //adb_init(path_multirom);
     run_multirom();
-    adb_quit();
+    //adb_quit();
     return 0;
 }
 
