diff --git a/kexec/arch/arm/kexec-zImage-arm.c b/kexec/arch/arm/kexec-zImage-arm.c
index 9c7c330..832f2c9 100644
--- a/kexec/arch/arm/kexec-zImage-arm.c
+++ b/kexec/arch/arm/kexec-zImage-arm.c
@@ -16,6 +16,7 @@
 #include <arch/options.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <linux/fs.h>
 #include <fcntl.h>
 #include <libfdt.h>
 #include "../../kexec.h"
@@ -336,6 +337,56 @@ static int load_dtb_image(const char *path, char **dtb_img, off_t *dtb_img_len)
 	return 1;
 }
 
+static int load_dtb_partition(const char *path, char **dtb_img, off_t *dtb_img_len)
+{
+	int fd;
+	struct stat info;
+	char buff[4];
+	char *img = NULL;
+	off_t size;
+
+	if(stat(path, &info) < 0)
+	{
+		fprintf(stderr, "DTB: Failed to stat() dtb partition %s\n", path);
+		return 0;
+	}
+
+	if(info.st_size <= 2048 && !S_ISBLK(info.st_mode))
+	{
+		fprintf(stderr, "DTB: invalid dtb partition (too small) %s\n", path);
+		return 0;
+	}
+
+	fd = open(path, O_RDONLY);
+	if(fd < 0)
+	{
+		fprintf(stderr, "DTB: Failed to open dtb partition %s\n", path);
+		return 0;
+	}
+
+	if (S_ISBLK(info.st_mode))
+	{
+		ioctl(fd, BLKGETSIZE, &size);
+		size *= 512;
+		printf("DTB: partition size = %d\n", size);
+	}
+
+	img = xmalloc(size);
+	if(read(fd, img, size) != size)
+	{
+		fprintf(stderr, "Failed to read %lld bytes from dtb partition %s\n", size, path);
+		free(img);
+		close(fd);
+		return 0;
+	}
+
+	*dtb_img = img;
+	*dtb_img_len = size;
+
+	close(fd);
+	return 1;
+}
+
 int zImage_arm_load(int argc, char **argv, const char *buf, off_t len,
 	struct kexec_info *info)
 {
@@ -525,12 +576,17 @@ int zImage_arm_load(int argc, char **argv, const char *buf, off_t len,
 			printf("DTB: Using DTB from file %s\n", dtb_file);
 			free_dtb_img = 1;
 		}
+		else if (get_appended_dtb(buf, len, &dtb_img, &dtb_img_len))
+		{
+			printf("DTB: Using DTB appended to zImage\n");
+		}
 		else
 		{
-			if(!get_appended_dtb(buf, len, &dtb_img, &dtb_img_len))
+			if(!load_dtb_partition("/dev/block/platform/sdhci-tegra.3/by-name/DTB", &dtb_img, &dtb_img_len))
 				return -1;
 
-			printf("DTB: Using DTB appended to zImage\n");
+			printf("DTB: Using DTB partition /dev/block/platform/sdhci-tegra.3/by-name/DTB\n");
+			free_dtb_img = 1;
 		}
 
 		choose_res = (mach->choose_dtb)(dtb_img, dtb_img_len, &dtb_buf, &dtb_length);
