diff --git a/core/tasks/kernel.mk b/core/tasks/kernel.mk
index af1bf55..b8c8a42 100644
--- a/core/tasks/kernel.mk
+++ b/core/tasks/kernel.mk
@@ -155,6 +155,22 @@ ifeq ($(TARGET_ARCH),arm)
     ccache = 
 endif
 
+ifeq ($(TARGET_ARCH),arm64)
+ifneq ($(KERNEL_TOOLCHAIN_PREFIX),)
+    ifneq ($(USE_CCACHE),)
+        ccache := $(ANDROID_BUILD_TOP)/prebuilts/misc/$(HOST_PREBUILT_TAG)/ccache/ccache
+        # Check that the executable is here.
+        ccache := $(strip $(wildcard $(ccache)))
+    endif
+    ifeq ($(HOST_OS),darwin)
+        ARM_CROSS_COMPILE:=CROSS_COMPILE="$(ccache) $(ANDROID_BUILD_TOP)/prebuilts/gcc/darwin-x86/aarch64/aarch64-linux-android-$(TARGET_GCC_VERSION)/bin/$(KERNEL_TOOLCHAIN_PREFIX)"
+    else
+        ARM_CROSS_COMPILE:=CROSS_COMPILE="$(ccache) $(ANDROID_BUILD_TOP)/prebuilts/gcc/linux-x86/aarch64/aarch64-linux-android-$(TARGET_GCC_VERSION)/bin/$(KERNEL_TOOLCHAIN_PREFIX)"
+    endif
+    ccache =
+endif
+endif
+
 ifeq ($(HOST_OS),darwin)
   MAKE_FLAGS := C_INCLUDE_PATH=$(ANDROID_BUILD_TOP)/external/elfutils/libelf
 endif
