diff --git a/minuitwrp/Android.mk b/minuitwrp/Android.mk
index be19117..45f0cf4 100644
--- a/minuitwrp/Android.mk
+++ b/minuitwrp/Android.mk
@@ -157,7 +157,7 @@ ifneq ($(TW_THEME_LANDSCAPE),)
   LOCAL_CFLAGS += -DTW_HAS_LANDSCAPE
 endif
 
-LOCAL_SHARED_LIBRARIES += libz libc libcutils libjpeg libpng
+LOCAL_SHARED_LIBRARIES += libz libc libcutils libjpeg libpng libutils
 LOCAL_STATIC_LIBRARIES += libpixelflinger_static
 LOCAL_MODULE_TAGS := eng
 LOCAL_MODULE := libminuitwrp
