diff --git a/partition.cpp b/partition.cpp
index 11ad3d1..5202667 100644
--- a/partition.cpp
+++ b/partition.cpp
@@ -1041,6 +1041,16 @@ bool TWPartition::Mount(bool Display_Error) {
 	if (Mount_Read_Only)
 		flags |= MS_RDONLY;
 
+	if (Current_File_System == "ntfs" && TWFunc::Path_Exists("/sbin/ntfs-3g") && !mounted) {
+		string cmd = "/sbin/ntfs-3g " + Actual_Block_Device + " " + Mount_Point;
+		LOGINFO("cmd: %s\n", cmd.c_str());
+		string result;
+		if (TWFunc::Exec_Cmd(cmd, result) != 0) {
+			LOGINFO("ntfs-3g failed to mount with result '%s', trying kernels ntfs\n", result.c_str());
+		} else {
+			mounted = 1;
+		}
+	}
 	if (Fstab_File_System == "yaffs2" && !mounted) {
 		// mount an MTD partition as a YAFFS2 filesystem.
 		flags = MS_NOATIME | MS_NODEV | MS_NODIRATIME;
